File: 257-binary-tree-paths/257.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

/**
 * Definition for a binary tree node.
 **/
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution
{
public:
    vector<string> binaryTreePaths(TreeNode *root)
    {
        vector<string> res;
        dfs(root, "", res);
        return res;
    }

    void dfs(TreeNode *root, string path, vector<string> &res)
    {
        if (root != nullptr)
        {
            path += to_string(root->val);
            if (!root->left && !root->right)
            {
                res.push_back(path);
            }
            else
            {
                path += "->";
                dfs(root->left, path, res);
                dfs(root->right, path, res);
            }
        }
    }
};

int main(int argc, char **args)
{
    Solution sol = Solution();
    return 0;
}

File: shoope/257-binary-tree-paths/257.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        def dfs(node: Optional[TreeNode], path: str):
            if node:
                path += str(node.val)
                if not node.left and not node.right:
                    paths.append(path)
                else:
                    path += '->'
                    dfs(node.left, path)
                    dfs(node.right, path)

        paths = []
        dfs(root, '')
        return paths


if __name__ == '__main__':
    sol = Solution()
